src/Documentation/DocumentationGenerator.php

```diff
@@ -0,0 +1,416 @@
+<?php
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Documentation;
+
+use PhpCsFixer\AbstractFixer;
+use PhpCsFixer\Console\Command\HelpCommand;
+use PhpCsFixer\Diff\GeckoPackages\DiffOutputBuilder\UnifiedDiffOutputBuilder;
+use PhpCsFixer\Diff\v2_0\Differ;
+use PhpCsFixer\Fixer\Basic\Psr0Fixer;
+use PhpCsFixer\Fixer\ConfigurableFixerInterface;
+use PhpCsFixer\Fixer\ConfigurationDefinitionFixerInterface;
+use PhpCsFixer\Fixer\DefinedFixerInterface;
+use PhpCsFixer\Fixer\DeprecatedFixerInterface;
+use PhpCsFixer\Fixer\FixerInterface;
+use PhpCsFixer\FixerConfiguration\AliasedFixerOption;
+use PhpCsFixer\FixerConfiguration\AllowedValueSubset;
+use PhpCsFixer\FixerConfiguration\DeprecatedFixerOptionInterface;
+use PhpCsFixer\FixerDefinition\CodeSampleInterface;
+use PhpCsFixer\FixerDefinition\FileSpecificCodeSampleInterface;
+use PhpCsFixer\FixerDefinition\VersionSpecificCodeSampleInterface;
+use PhpCsFixer\Preg;
+use PhpCsFixer\RuleSet;
+use PhpCsFixer\StdinFileInfo;
+use PhpCsFixer\Tokenizer\Tokens;
+use PhpCsFixer\Utils;
+
+class DocumentationGenerator
+{
+    /**
+     * @var Differ
+     */
+    private $differ;
+
+    private $path;
+
+    public function __construct()
+    {
+        $this->differ = new Differ(new UnifiedDiffOutputBuilder([
+            'fromFile' => 'Original',
+            'toFile' => 'New',
+        ]));
+
+        $this->path = \dirname(\dirname(__DIR__)).'/doc/rules';
+    }
+
+    /**
+     * @return string
+     */
+    public function getFixersDocumentationDirectoryPath()
+    {
+        return $this->path;
+    }
+
+    /**
+     * @return string
+     */
+    public function getFixersDocumentationIndexFilePath()
+    {
+        return "{$this->path}/index.rst";
+    }
+
+    /**
+     * @param AbstractFixer[] $fixers
+     *
+     * @return string
+     */
+    public function generateFixersDocumentationIndex(array $fixers)
+    {
+        $overrideGroups = [
+            'PhpUnit' => 'PHPUnit',
+            'PhpTag' => 'PHP Tag',
+            'Phpdoc' => 'PHPDoc',
+        ];
+
+        usort($fixers, function (FixerInterface $a, FixerInterface $b) {
+            return strcmp(\get_class($a), \get_class($b));
+        });
+
+        $documentation = <<<'RST'
+=======================
+List of Available Rules
+=======================
+RST;
+
+        $currentGroup = null;
```

Comment:
What is the profit of having the groups? Would that help anyhow to someone looking for a rule?

Reply:
I do find reading the list with groups easier: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/ff12f49c185da6eac090595519e5902b2305283d/doc/rules/index.rst